pkg/function/restore_csi_snapshot_test.go

```diff
@@ -0,0 +1,168 @@
+var _ = Suite(&RestoreCSISnapshotTestSuite{})
+
+func (testSuite *RestoreCSISnapshotTestSuite) SetUpSuite(c *C) {
+	testSuite.volumeSnapshotClass = RestoreCSISnapshotSnapshotClass
+	testSuite.storageClass = RestoreCSISnapshotStorageClass
+	testSuite.pvcName = RestoreCSISnapshotOriginalPVCName
+	testSuite.newPVCName = RestoreCSISnapshotNewPVCName
+	testSuite.snapName = RestoreCSISnapshotSnapshotName
+	testSuite.namespace = RestoreCSISnapshotTestNamespace
+}
+
+func (testSuite *RestoreCSISnapshotTestSuite) TestRestoreCSISnapshot(c *C) {
+	for _, apiResourceList := range []*metav1.APIResourceList{
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "VolumeSnapshot",
+				APIVersion: "v1alpha1",
+			},
+			GroupVersion: "snapshot.storage.k8s.io/v1alpha1",
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "VolumeSnapshot",
+				APIVersion: "v1beta1",
+			},
+			GroupVersion: "snapshot.storage.k8s.io/v1beta1",
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "VolumeSnapshot",
+				APIVersion: "v1",
+			},
+			GroupVersion: "snapshot.storage.k8s.io/v1",
+		},
+	} {
+		fakeCli := fake.NewSimpleClientset()
+		fakeCli.Resources = []*metav1.APIResourceList{apiResourceList}
+
+		_, err := fakeCli.CoreV1().Namespaces().Create(context.TODO(), &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testSuite.namespace}}, metav1.CreateOptions{})
+		c.Assert(err, IsNil)
+
+		scheme := runtime.NewScheme()
+		fakeSnapshotter, err := snapshot.NewSnapshotter(fakeCli, dynfake.NewSimpleDynamicClient(scheme))
+		c.Assert(err, IsNil)
+
+		originalPVC := getOriginalPVCManifest(testSuite.pvcName, testSuite.storageClass)
+		createPVC(c, testSuite.namespace, originalPVC, fakeCli)
+
+		err = fakeSnapshotter.Create(context.Background(), testSuite.snapName, testSuite.namespace, testSuite.pvcName, &testSuite.volumeSnapshotClass, false, nil)
+		c.Assert(err, IsNil)
+
+		vs, err := fakeSnapshotter.Get(context.Background(), testSuite.snapName, testSuite.namespace)
+		c.Assert(err, IsNil)
+		c.Assert(vs.Name, Equals, testSuite.snapName)
+
+		restoreArgs := restoreCSISnapshotArgs{
+			Name:         testSuite.snapName,
+			PVC:          testSuite.newPVCName,
+			Namespace:    testSuite.namespace,
+			StorageClass: testSuite.storageClass,
+			RestoreSize:  originalPVC.Spec.Resources.Requests.Storage(),
+			Labels:       nil,
+		}
+
+		var invalidVolumeMode v1.PersistentVolumeMode = "test"
+		restoreArgs.VolumeMode = invalidVolumeMode
+		err = validateVolumeModeArg(restoreArgs)
+		c.Assert(err, NotNil)
+
+		var invalidAccessMode []v1.PersistentVolumeAccessMode = []v1.PersistentVolumeAccessMode{"test"}
+		restoreArgs.AccessModes = invalidAccessMode
+		err = validateVolumeAccessModesArg(restoreArgs)
+		c.Assert(err, NotNil)
+
+		restoreArgs.VolumeMode = *originalPVC.Spec.VolumeMode
+		err = validateVolumeModeArg(restoreArgs)
+		c.Assert(err, IsNil)
+
+		restoreArgs.AccessModes = originalPVC.Spec.AccessModes
+		err = validateVolumeModeArg(restoreArgs)
+		c.Assert(err, IsNil)
```

Comment:
Let's write another test that would be table driven to these utility functions.

Reply:
@viveksinghggits Done. Please have a look.